src/dmd/inline.d

```diff
@@ -123,69 +123,59 @@ public:
     override void visit(CompoundStatement s)
     {
         //printf("CompoundStatement.doInlineAs!%s() %d\n", Result.stringof.ptr, s.statements.dim);
+        visitCompoundStatement(s, 0);
+    }
+
+    private void visitCompoundStatement(CompoundStatement s, size_t n)
+    {
         static if (asStatements)
         {
             auto as = new Statements();
             as.reserve(s.statements.dim);
         }
 
-        foreach (i, sx; *s.statements)
+        foreach (i; n .. s.statements.dim)
         {
+            auto sx = (*s.statements)[i];
             if (!sx)
                 continue;
+
             static if (asStatements)
             {
                 as.push(doInlineAs!Statement(sx, ids));
+                if (ids.foundReturn)
+                    break;
             }
             else
             {
-                /* Specifically allow:
-                 *  if (condition)
-                 *      return exp1;
-                 *  return exp2;
-                 */
-                IfStatement ifs;
-                Statement s3;
-                if ((ifs = sx.isIfStatement()) !is null &&
-                    ifs.ifbody &&
-                    ifs.ifbody.isReturnStatement() &&
-                    !ifs.elsebody &&
-                    i + 1 < s.statements.dim &&
-                    (s3 = (*s.statements)[i + 1]) !is null &&
-                    s3.isReturnStatement()
-                   )
+                if (auto ifs = sx.isIfStatement())
                 {
-                    /* Rewrite as ?:
-                     */
-                    auto econd = doInlineAs!Expression(ifs.condition, ids);
-                    assert(econd);
-                    auto e1 = doInlineAs!Expression(ifs.ifbody, ids);
-                    assert(ids.foundReturn);
-                    auto e2 = doInlineAs!Expression(s3, ids);
-
-                    Expression e = new CondExp(econd.loc, econd, e1, e2);
-                    e.type = e1.type;
-                    if (e.type.ty == Ttuple)
+                    scope DoInlineAs!Result isv = new DoInlineAs!Result(ids);
+                    bool subseqIsVisited = false;
```

Comment:
What does `subseq` refer ti>

Reply:
According to Kenji:
> Subsequent statements of IfStatement.

Judging from the control flow, whilst inlining all CompoundStatements from `n .. dim` (starting at 0), the context is switched over to visitIfStatement if we come across an IfStatement.  It is passed a delegate that switches it back to visitCompoundStatement to inline the remaining statements that follow.

It looks like it will remain false if there were returns found in both branches of the IfStatement.
```
if (foo)
  return 1;
else
  return 2;
// subsequent statements are unreachable.
```

Reply:
If only one of the branches returns, then subsequent statements are moved into the other (the one that doesn't return).

So in the example I gave:
```
pragma(inline, true)
{
    int foo3(int v)
    {
        return bar3(2 * v);
    }

    int bar3(int a)
    {
        if (a > 0)
            a = 1;
        else
            return baz3(a);
        return a;
    }
}
```
Function `foo3` would inline `bar3` as:
```
    int foo3(int v)
    {
        int a = 2 * v;
        if (a > 0)
        {
            a = 1;
            return a;
        }
        else
            return baz3(a);
    }
```
Actual dump from `-vcg-ast` is a bit more succinct however.
```
    int foo3(int v)
    {
        return (int a = 2 * v;) , a > 0 ? a = 1 , a : baz3(a);
    }
```

Reply:
When I ask for clarification on something, I'm implicitly suggesting it needs more explanation in the comments, not just on github.

Reply:
They are notes for me to look at later.  I'm trying to understand the original just as much as you are.

Reply:
> They are notes for me to look at later.

As I'm about to pick this up again and I for sure would have forgotten had I not noted it down. 😉